src/smc-webapp/widget-markdown-input/main.tsx

```diff
@@ -0,0 +1,265 @@
+// 3rd Party Libraries
+import * as markdown from "../markdown";
+import { Button, ButtonToolbar, FormControl, FormGroup } from "react-bootstrap";
+
+// Internal Libraries
+import { Component, React, ReactDOM, rclass, redux, rtypes } from "../app-framework";
+
+// Sibling Libraries
+import * as info from "./info";
+import { MarkdownWidgetActions } from "./actions";
+import { MarkdownWidgetStore, MarkdownWidgetStoreState } from "./store";
+
+export function init(): void {
+  if (redux.hasActions(info.name)) {
+    return;
+  }
+  redux.createStore(info.name, MarkdownWidgetStore);
+  redux.createActions(info.name, MarkdownWidgetActions);
+};
+
+interface ReactProps {
+  autoFocus: boolean;
+  persist_id: string; // A unique id to identify the input. Required if you want automatic persistence
+  attach_to: string; // Removes record when given store name is destroyed. Only use with persist_id
+  default_value: string;
+  editing: boolean; // Used to control the edit/display state. CANNOT be used with persist_id
+  save_disabled: boolean; // Used to control the save button
+  on_change: (value: string) => any; // called with the new value when the value while editing changes
+  on_save: (value: string) => any; // called when saving from editing and switching back
+  on_edit: (value: string) => any; // called when editing starts
+  on_cancel: (value: string) => any; // called when cancel button clicked
+  rows: number;
+  placeholder: string;
+  rendered_style: object;
+  hide_edit_button: boolean;
+}
+
+interface ReduxProps {
+  open_inputs: Map<any, any>;
+}
+
+interface MarkdownInputState {
+  editing: boolean;
+  value: string;
+}
+
+class MarkdownInput0 extends Component<ReactProps & ReduxProps, MarkdownInputState> {
+  displayName: "WidgetMarkdownInput";
+
+  constructor(props) {
+    super(props);
+    this.state = this.getInitialState();
+    this.getInitialState = this.getInitialState.bind(this);
+    this.componentDidMount = this.componentDidMount.bind(this);
```

Comment:
This binding stuff is definitely not needed.  Delete it.  Overall this file seems like a good *model*, since it shows how to use reduxProps properly, and looks clean.  So let's do it right.